src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala

```diff
@@ -0,0 +1,184 @@
+      stop(self)
+
+    case Failure(e) =>
+      // An error occurred, let the supervisor handle it
+      logger.error(s"Error occurred monitoring cluster ${cluster.googleProject}/${cluster.clusterName}", e)
+      throw e
+  }
+
+  private def scheduleMonitorPass: Unit = {
+    system.scheduler.scheduleOnce(monitorConfig.pollPeriod, self, QueryForCluster)
+  }
+
+  /**
+    * Handles a dataproc cluster which is not ready yet. We don't take any action, just
+    * schedule another monitor pass.
+    * @param status the ClusterStatus from Google
+    * @return ScheduleMonitorPass
+    */
+  private def handleNotReadyCluster(status: ClusterStatus): Future[ClusterMonitorMessage] = {
+    logger.info(s"Cluster ${cluster.googleProject}/${cluster.clusterName} is not ready yet ($status). Checking again in ${monitorConfig.pollPeriod.toString}.")
+    Future.successful(ScheduleMonitorPass)
+  }
+
+  /**
+    * Handles a dataproc cluster which is ready. We update the status and IP in the database,
+    * then shut down this actor.
+    * @param publicIp the cluster public IP, according to Google
+    * @return ShutdownActor
+    */
+  private def handleReadyCluster(publicIp: String): Future[ClusterMonitorMessage] = {
+    logger.info(s"Cluster ${cluster.googleProject}/${cluster.clusterName} is ready for use!")
+    dbRef.inTransaction { dataAccess =>
+      dataAccess.clusterQuery.updateClusterStatus(cluster.googleId, ClusterStatus.Running) andThen
+        dataAccess.clusterQuery.updateIpByGoogleId(cluster.googleId, publicIp)
+    }.map(_ => ShutdownActor())
+  }
+
+  /**
+    * Handles a dataproc cluster which has failed. If it is a recoverable error, then we will
+    * try to delete and recreate the cluster. Otherwise, just set the status to Error in the
+    * database and stop monitoring the cluster.
+    * @param errorDetails cluster error details from Google
+    * @return ShutdownActor
+    */
+  private def handleFailedCluster(errorDetails: ClusterErrorDetails): Future[ClusterMonitorMessage] = {
+    if (shouldRecreateCluster(errorDetails.code, errorDetails.message)) {
+      // Delete the cluster in Google and update the database record to Deleting.
+      // Then shutdown this actor, but register a callback message to the supervisor
+      // telling it to recreate the cluster.
+      logger.info(s"Cluster ${cluster.googleProject}/${cluster.clusterName} is in an error state with $errorDetails. Attempting to delete and recreate...")
+      for {
+        _ <- gdDAO.deleteCluster(cluster.googleProject, cluster.clusterName)
+        _ <- dbRef.inTransaction { _.clusterQuery.deleteCluster(cluster.googleId) }
+      } yield ShutdownActor(Some(ClusterDeleted(cluster, recreate = true)))
+    } else {
+      // Update the database record to Error and shutdown this actor.
+      logger.warn(s"Cluster ${cluster.googleProject}/${cluster.clusterName} is in an error state with $errorDetails'. Unable to recreate cluster.")
```

Comment:
i think you also want to delete the cluster in google here too (i've definitely seen clusters in error states just sitting there)

Reply:
Hm, ok. So for failed clusters we'll delete them no matter what, and potentially try to recreate if we think it's recoverable.